include/etcpal/cpp/common.h

```diff
@@ -48,4 +55,21 @@
 
 /// @}
 
-#endif // ETCPAL_CPP_COMMON_H_
+namespace etcpal
+{
+/// \cond detail
+
+namespace detail
+{
+enum class enabler
+{
+};
+
+#define ETCPAL_ENABLE_IF_ARG(...) typename std::enable_if<(__VA_ARGS__)>::type* = nullptr
+#define ETCPAL_ENABLE_IF_TEMPLATE(...) typename = typename std::enable_if<(__VA_ARGS__), detail::enabler>::type
```

Comment:
I want to understand these better.  What exactly are you doing?

Reply:
These are SFINAE constructs that I don't fully understand copied from https://github.com/martinmoene/expected-lite. They are for removing overloads from overload resolution based on certain conditions.